Fix a bug that conversion result is wrong

komainu8 opened the pull request.

komainu8 commented:
@kenhys Could you review this PR, please?

komainu8 commented:
@kou Could you review this PR, please?

kou commented on lib/groonga/command/format/elasticsearch.rb:
```diff
@@ -57,27 +59,27 @@ def json(options={})
               end
               components << JSON.generate(header)
+            when :columns
+              value.split(",").each do |column_name|
```
We should use `split(/\s*,\s*)` because "_key, column" is a valid syntax.

BTW, why do you split the command object to name and arguments? `Groonga::Command::Load#columns` parses raw columns value.

komainu8 replied:
Just imitated the implementation of `Groonga:Command:Format:Command`, there is no more specific reason for split the command object to name and arguments.

Therefore, I use `Groonga::Command::Load`.

kou commented on lib/groonga/command/format/elasticsearch.rb:
```diff
@@ -57,27 +59,27 @@ def json(options={})
               else
                 record.each do |load_value|
                   components << JSON.generate(load_value)
+                  components << JSON.generate(header) if load_value != record.last
```
This is weird.
If there is no "values", can this code generate correct output?

```ruby
      class Elasticsearch
        def initialize(command)
          @command = command
        end

        def json(options={})
          return nil unless @command.is_a?(Load)

          components = []
          action = build_action
          each_source do |source|
            components << JSON.generate(action)
            components << JSON.generate(source)
          end
          components.join("\n")
        end

        private
        def elasticsearch_version
          options[:version] || 5
        end

        def build_action
          index_name = @command.table.downcase
          case elasticsearch_version
          when 5, 6
            {
              "index" => {
                "_index" => index_name,
                "_type" => "groonga",
              }
            }
          when 7
            {
              "index" => {
                "_index" => index_name,
                "_type"=>"_doc",
              }
            }
          else
            {
              "index" => {
                "_index" => index_name,
              }
            }
          end
        end

        def each_source
          columns = @command.columns
          values = @command.values || []
          values.each do |value|
            if value.is_a?(::Array)
              if columns.nil?
                columns = value
              else
                yield(Hash[columns.zip(value)])
              end
            else
              yield(value)
            end
          end
        end
```

komainu8 replied:
Thank you for your suggestion.

I've changed your suggestion code a little bit.
Because `@command.columns` and `@command.values` are nil, I used `@command[:columns]` and `@command[:values]` instead.

kou commented on test/command/test-base.rb:
```diff
@@ -138,9 +138,9 @@ def test_select_command
+                                          :columns => "_key,title",
                                           :values => <<-VALUES)
   [
-  ["_key","title"],
```
Could you add a test that doesn't use `:columns` and uses columns in `:values`.?

komainu8 commented:
@kou Thank you for your comments.
I've fixed them.

I've changed your suggestion code a little bit.
Because `@command.columns` and `@command.values` are nil, I used `@command[:columns]` and `@command[:values]` instead.

kou commented:
Really? It works well with the following change:

```diff
diff --git a/lib/groonga/command/format/elasticsearch.rb b/lib/groonga/command/format/elasticsearch.rb
index 9b513d1..6d5eb0b 100644
--- a/lib/groonga/command/format/elasticsearch.rb
+++ b/lib/groonga/command/format/elasticsearch.rb
@@ -40,7 +40,7 @@ module Groonga
         end
 
         def build_action(options)
-          index_name = @command[:table].downcase
+          index_name = @command.table.downcase
           case elasticsearch_version(options)
           when 5, 6
             {
@@ -66,10 +66,8 @@ module Groonga
         end
 
         def each_source
-          unless @command[:columns].nil?
-            columns = @command[:columns].split(/\s*,\s*/)
-          end
-          values = JSON.parse(@command[:values]) || []
+          columns = @command.columns
+          values = @command.values || []
           values.each do |value|
             if value.is_a?(::Array)
               if columns.nil?
```

```console
% test/run-test.rb
Loaded suite test
Started
................................................................................
................................................................................
................................................................................
..........................................................
Finished in 0.085880992 seconds.
--------------------------------------------------------------------------------
298 tests, 298 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
--------------------------------------------------------------------------------
3469.92 tests/s, 3469.92 assertions/s
```

I think that you used `Groonga::Command::Base` instead of `Groonga::Command::Load`.

komainu8 commented:
@kou 
I'm sorry. I was mistaken about it...
Certainly, the above change has worked well.

Therefore, I've fixed it.
